lib/chef/knife.rb

```diff
@@ -343,8 +343,15 @@ def cli_keys
     # Chef::Config[:knife] would break the defaults in the cli that we would otherwise
     # overwrite.
     def config_file_settings
+      @key_sources = { cli: [], config: [] }
       cli_keys.each_with_object({}) do |key, memo|
-        memo[key] = Chef::Config[:knife][key] if Chef::Config[:knife].key?(key)
+        if config.key?(key)
+          @key_sources[:cli] << key
+        end
+        if Chef::Config[:knife].key?(key)
+          @key_sources[:config] << key
+          memo[key] = Chef::Config[:knife][key] if Chef::Config[:knife].key?(key)
```

Comment:
Do we need if Chef::Config[:knife].key?(key) check here again as we are already checking that at line 351 ?